File: android/core/src/commonMain/kotlin/com/hexstacker/core/render/Theme.kt
package com.hexstacker.core.render

/**
 * Palettes, style tiers, and theme tokens, ported from `public/shared/theme.js`.
 * Colors held as [Rgb], parity-checked against the JS hex strings.
 */
object Theme {

    val partyPalette: List<Rgb> = listOf(
        "#FF6B6B", "#4ECDC4", "#FFE066", "#A78BFA",
        "#7BED6F", "#F178D8", "#5B7FFF", "#FF8C42",
    ).map { Rgb.fromHex(it)!! }

    /** grid cell value -> color. 1..6 piece types, 9 garbage. Index 0 (empty) omitted. */
    val pieceColors: Map<Int, Rgb> = mapOf(
        1 to partyPalette[0], // #FF6B6B I3 red
        2 to partyPalette[1], // #4ECDC4 V3 teal
        3 to partyPalette[2], // #FFE066 T3 honey
        4 to partyPalette[3], // #A78BFA o  violet
        5 to partyPalette[4], // #7BED6F d  mint
        6 to partyPalette[5], // #F178D8 b  magenta
        9 to Rgb.fromHex("#808080")!!, // garbage gray (off-palette)
    )

    /** Reordered to a visible spectrum (NOT palette order); the remap is load-bearing. */
    val playerColors: List<Rgb> = listOf(
        partyPalette[0], // 0 Red
        partyPalette[7], // 1 Tangerine
        partyPalette[2], // 2 Honey
        partyPalette[4], // 3 Mint
        partyPalette[1], // 4 Teal
        partyPalette[6], // 5 Indigo
        partyPalette[3], // 6 Violet
        partyPalette[5], // 7 Magenta
    )

    /**
     * Out-of-range slots fall back to the index-0 color (Red), matching the board/UI
     * renderers' `PLAYER_COLORS[playerIndex] || PLAYER_COLORS[0]` guard
     * (BoardRenderer.js:25, UIRenderer.js:76); this helper feeds the native board/UI
     * renderers, so it mirrors them, not the lobby list's white `|| '#fff'` fallback
     * (DisplayUI.js:175). Latent today: MAX_PLAYERS == playerColors.size (8), so a slot
     * is always in range; aligned to avoid a future footgun.
     */
    fun playerColor(slot: Int): Rgb = playerColors.getOrElse(slot) { playerColors[0] }

    val bgPrimary = Rgb.fromHex("#1E1A2B")!!
    val bgSecondary = Rgb.fromHex("#181421")!!
    val bgBoard = Rgb.fromHex("#15121F")!!

    /**
     * Warm-paper hairline for socket rims and ring strokes — mirrors the
     * rgba(255,248,236,X) --border family in theme.css.
     */
    val hairline = Rgb.fromHex("#FFF8EC")!!

    /**
     * Cream outline pulse (== text primary #F7F1E8) — extends the "cream =
     * clear-related" vocabulary shared with the clear preview and clear glow.
     */
    val nearClear = Rgb.fromHex("#F7F1E8")!!

    /** Warm cream text primary (#F7F1E8) — canvas twin of --text-primary. */
    val textPrimary = nearClear

    object Opacity {
        const val faint = 0.04
        const val tint = 0.06
        const val boardTint = 0.12
        const val subtle = 0.08
        const val muted = 0.10
        const val grid = 0.18
        const val soft = 0.15
        const val hairline = 0.12 // socket-rim hairline strokes (× color hairline)
        const val highlight = 0.16
        const val shadow = 0.25
        const val label = 0.6
        const val wall = 0.5 // board wall stroke (player color)
        const val overlay = 0.75
        const val panel = 0.9
    }

    object Stroke {
        const val grid = 0.03
        const val border = 0.04
        const val ghost = 0.05
    }

    object Size {
        const val panelWidth = 4.5
        const val panelGap = 0.25
        const val canvasPad = 5.0
        const val blockGap = 0.03
        const val tvOverscan = 0.05 // TV title-safe margin per edge (Google/Apple ~5% overscan)
    }

    // Mirrors web THEME.font (public/shared/Theme.js): cellScale + minPx.
    object Font {
        const val nameScale = 0.9
        const val labelScale = 0.48
        const val miniScale = 0.6
        const val nameMinPx = 24.0
        const val labelMinPx = 14.0
    }

    enum class StyleTier { NORMAL, PILLOW, NEON_FLAT }

    /** getStyleTier: Lv 1-5 NORMAL, 6-10 PILLOW, 11+ NEON_FLAT. */
    fun styleTier(level: Int): StyleTier = when {
        level >= 11 -> StyleTier.NEON_FLAT
        level >= 6 -> StyleTier.PILLOW
        else -> StyleTier.NORMAL
    }

    /** The JS string token a tier maps to, for parity comparison. */
    fun styleTierToken(t: StyleTier): String = when (t) {
        StyleTier.NORMAL -> "normal"
        StyleTier.PILLOW -> "pillow"
        StyleTier.NEON_FLAT -> "neonFlat"
    }
}

File: appletv/Sources/HexStackerKit/Render/Theme.swift
import Foundation

// Visual theme values mirrored from public/shared/theme.js. RGB tuples are kept
// platform-agnostic (the tvOS layer maps them to UIColor); this lets the values
// be unit-checked on macOS.

public enum Theme {

    // Cell value (grid / typeId) -> color. 1...6 are piece types, 9 is garbage.
    public static let pieceColors: [Int: RGB] = [
        1: RGB(0xFF, 0x6B, 0x6B),  // I3 red
        2: RGB(0x4E, 0xCD, 0xC4),  // V3 teal
        3: RGB(0xFF, 0xE0, 0x66),  // T3 honey
        4: RGB(0xA7, 0x8B, 0xFA),  // o  violet
        5: RGB(0x7B, 0xED, 0x6F),  // d  mint
        6: RGB(0xF1, 0x78, 0xD8),  // b  magenta
        9: RGB(0x80, 0x80, 0x80),  // garbage gray
    ]

    // Per-player accent palette (color slot -> color), spectrum-ordered.
    public static let playerColors: [RGB] = [
        RGB(0xFF, 0x6B, 0x6B),  // 0 red
        RGB(0xFF, 0x8C, 0x42),  // 1 tangerine
        RGB(0xFF, 0xE0, 0x66),  // 2 honey
        RGB(0x7B, 0xED, 0x6F),  // 3 mint
        RGB(0x4E, 0xCD, 0xC4),  // 4 teal
        RGB(0x5B, 0x7F, 0xFF),  // 5 indigo
        RGB(0xA7, 0x8B, 0xFA),  // 6 violet
        RGB(0xF1, 0x78, 0xD8),  // 7 magenta
    ]

    public static func playerColor(slot: Int) -> RGB {
        // Out-of-range falls back to slot 0 (red), matching the web's
        // `PLAYER_COLORS[idx] || PLAYER_COLORS[0]` (e.g. UIRenderer.js).
        guard slot >= 0, slot < playerColors.count else { return playerColors[0] }
        return playerColors[slot]
    }

    // Background colors.
    public static let bgPrimary = RGB(0x1E, 0x1A, 0x2B)    // main canvas
    public static let bgSecondary = RGB(0x18, 0x14, 0x21)  // deeper plum panel bg
    public static let bgBoard = RGB(0x15, 0x12, 0x1F)      // flat recessed well fill / socket base
    // Warm-paper hairline for socket rims and ring strokes — mirrors the
    // rgba(255,248,236,X) --border family in theme.css (THEME.color.hairline).
    public static let hairline = RGB(0xFF, 0xF8, 0xEC)
    // Cream outline pulse (== text.primary) — extends the "cream =
    // clear-related" vocabulary shared with the clear preview and clear glow.
    public static let nearClear = RGB(0xF7, 0xF1, 0xE8)

    // Opacity tokens.
    public enum Opacity {
        public static let boardTint = 0.12
        public static let grid = 0.18
        public static let hairline = 0.12  // socket-rim hairline strokes (× hairline color)
        public static let wall = 0.5       // board wall stroke (player color)
        public static let highlight = 0.16
        public static let shadow = 0.25
        public static let overlay = 0.75   // dark overlays (board-area KO/disconnect dim)
        public static let subtle = 0.08
    }

    // Size / stroke tokens.
    public enum Size {
        public static let canvasPad = 5.0   // outer margin and inter-tile gap
        public static let blockGap = 0.03
    }
    public enum Stroke {
        public static let grid = 0.03
        public static let border = 0.04
        public static let ghost = 0.05
    }

    // MARK: Style tiers (driven by player level)

    public enum StyleTier: Equatable {
        case normal     // Lv 1-5: gradient + highlight/shadow bands
        case pillow     // Lv 6-10: rounded, radial gloss
        case neonFlat   // Lv 11+: dark fill + bright rim
    }

    public static func tier(forLevel level: Int) -> StyleTier {
        if level >= 11 { return .neonFlat }
        if level >= 6 { return .pillow }
        return .normal
    }
}

// MARK: - Multi-board layout (ported from DisplayUI.calculateLayout)

public struct BoardPlacement: Equatable {
    /// Board origin (top-left of bounding box), canvas Y-down, viewport space.
    public let originX: Double
    public let originY: Double
    public let seatIndex: Int
}

public struct BoardLayout: Equatable {
    public let cellSize: Double
    public let geometry: HexGeometry
    public let gridCols: Int
    public let gridRows: Int
    public let placements: [BoardPlacement]
}

public enum LayoutEngine {
    private static let padding = Theme.Size.canvasPad
    private static let totalCellsWide = Double(EngineConstants.cols) + 3 + 3   // board + side panels

    /// Height reserved above the board for the player-name label. Approximation:
    /// the web measures real glyph metrics here (DisplayUI.js measureHeight,
    /// ascent+descent of 'Mg' ≈ 0.9·nameSize), so native boards come out
    /// marginally smaller/lower for the same viewport. Layout is not part of the
    /// byte-parity contract; keep the two visually close if either side changes.
    public static func textHeight(_ cs: Double) -> Double {
        let nameSize = max(24.0, cs * 0.9)
        let nameGap = cs * 0.6
        return nameSize + nameGap
    }

    /// Largest integer cellSize that fits one board + name into a tile of the
    /// given grid over the viewport.
    public static func cellSizeFor(viewportW w: Double, viewportH h: Double,
                                   tileCols cols: Int, tileRows rows: Int) -> Double {
        let boardRowsUnits = HexGeometry(cellSize: 1).boardHeight
        let aw = (w - padding * Double(cols + 1)) / Double(cols)
        let ah = (h - padding * Double(rows + 1)) / Double(rows)
        var cs = (min(aw / totalCellsWide, ah / boardRowsUnits)).rounded(.down)
        while cs > 1 && cs * boardRowsUnits + textHeight(cs) > ah { cs -= 1 }
        return max(cs, 1)
    }

    /// Pick the tile grid (cols, rows) for n players, choosing whichever
    /// arrangement yields the larger cellSize at the breakpoints.
    public static func chooseGrid(playerCount n: Int, viewportW w: Double, viewportH h: Double) -> (cols: Int, rows: Int) {
        switch max(n, 1) {
        case 1: return (1, 1)
        case 2: return (2, 1)
        case 3: return (3, 1)
        case 4:
            return cellSizeFor(viewportW: w, viewportH: h, tileCols: 4, tileRows: 1)
                >= cellSizeFor(viewportW: w, viewportH: h, tileCols: 2, tileRows: 2) ? (4, 1) : (2, 2)
        case 5, 6:
            return cellSizeFor(viewportW: w, viewportH: h, tileCols: n, tileRows: 1)
                >= cellSizeFor(viewportW: w, viewportH: h, tileCols: 3, tileRows: 2) ? (n, 1) : (3, 2)
        default: // 7, 8 (and clamp above)
            let nn = min(n, EngineConstants.maxPlayers)
            return cellSizeFor(viewportW: w, viewportH: h, tileCols: nn, tileRows: 1)
                >= cellSizeFor(viewportW: w, viewportH: h, tileCols: 4, tileRows: 2) ? (nn, 1) : (4, 2)
        }
    }

    public static func layout(playerCount n: Int, viewportW w: Double, viewportH h: Double) -> BoardLayout {
        let (gridCols, gridRows) = chooseGrid(playerCount: n, viewportW: w, viewportH: h)
        let cellSize = cellSizeFor(viewportW: w, viewportH: h, tileCols: gridCols, tileRows: gridRows)
        let geo = HexGeometry(cellSize: cellSize)

        let maxSlots = gridCols * gridRows
        let cellAreaW = (w - padding * Double(gridCols + 1)) / Double(gridCols)
        let cellAreaH = (h - padding * Double(gridRows + 1)) / Double(gridRows)
        let nameArea = textHeight(cellSize)
        let totalContentH = geo.boardHeight + textHeight(cellSize)

        var placements: [BoardPlacement] = []
        for i in 0..<min(n, maxSlots) {
            let col = i % gridCols
            let row = i / gridCols
            let boardX = padding + Double(col) * (cellAreaW + padding) + (cellAreaW - geo.boardWidth) / 2
            let boardY = padding + Double(row) * (cellAreaH + padding)
                + (cellAreaH - totalContentH) / 2 + nameArea
            placements.append(BoardPlacement(originX: boardX, originY: boardY, seatIndex: i))
        }

        return BoardLayout(cellSize: cellSize, geometry: geo, gridCols: gridCols, gridRows: gridRows, placements: placements)
    }
}
